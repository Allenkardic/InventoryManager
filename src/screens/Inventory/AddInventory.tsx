import React from 'react';
import {ScrollView, StyleSheet, TextStyle, View} from 'react-native';
// components
import NavigationHeader from '../../components/NavigationHeader';
import CustomInput from '../../components/CustomInput';
import CustomButton from '../../components/CustomButton';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {StackModels} from '../../navigations/models';
// context
import {AuthUserContext} from '../../context/AuthUserContext';
import {InventoryListContext} from '../../context/InventoryContext';

import {
  stampId,
  checkIfItemNameExist,
  checkWordsCount,
} from '../../utils/constants';
import {COLORS, SPACING} from '../../utils/themes';

import {showMessage} from 'react-native-flash-message';
// form
import {useForm, Controller} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import * as yup from 'yup';

type Style = {
  formContainer: TextStyle;
  btnContainer: TextStyle;
};
// navigation type
type Props = NativeStackScreenProps<StackModels, 'AddInventory'>;

// form type
type FormValues = {
  inventoryName: string;
  stockCount: number;
  description: string;
  price: number;
};

function AddInventory({navigation}: Props) {
  const {authUserStore} = React.useContext(AuthUserContext);
  const {inventoryStore, setInventoryStore} =
    React.useContext(InventoryListContext);

  // filter inventory base on the user that created it
  const inventoryByUser = inventoryStore.filter(
    item => item.userEmail === authUserStore.email,
  );

  // validation schema
  const schema = yup.object().shape({
    inventoryName: yup
      .string()
      .required('Inventory name is required')
      .test(
        'inventoryName',
        'Inventory name already exists',
        (value: any) => checkIfItemNameExist(value, inventoryByUser) === false,
      ),
    stockCount: yup.number().required('Stock count left is required'),
    description: yup
      .string()
      .required('Brief descripting is required')
      .test(
        'description',
        'Description must not be less than 3 words',
        (value: any) => checkWordsCount(value) >= 3,
      ),
    price: yup.number().required('Price is required'),
  });

  // React Hook Form
  const {
    control,
    handleSubmit,
    formState: {errors},
  } = useForm<FormValues>({
    resolver: yupResolver(schema),
  });

  const onSubmit = (data: FormValues) => {
    const payload = {
      id: stampId(),
      userEmail: authUserStore.email,
      inventoryName: data.inventoryName,
      stockCount: data.stockCount,
      description: data.description,
      price: data.price,
    };

    const addNewPayload = [...inventoryStore, payload];
    // add data to inventory list
    setInventoryStore(addNewPayload);
    showMessage({
      type: 'success',
      message: 'Item added successfuly',
    });
    navigation.navigate('Inventory');
  };

  // console.log(checkIfItemNameExist('allen', inventoryStore), 'storesss');
  return (
    <>
      <View>
        <NavigationHeader
          onPressBackBtn={() => navigation.goBack()}
          title={'Add Inventory'}
          backgroundColor={COLORS.smokeWhite}
        />
      </View>
      <ScrollView>
        <View
          style={{
            ...styles.formContainer,
          }}>
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: book"
                label="Item name"
                backgroundColor={COLORS.lightWhite}
                error={errors.inventoryName?.message}
                marginBottom={SPACING.xsmall}
              />
            )}
            name="inventoryName"
          />
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 200"
                label="Price"
                backgroundColor={COLORS.lightWhite}
                error={errors.price?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="price"
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 10"
                label="Count"
                backgroundColor={COLORS.lightWhite}
                error={errors.stockCount?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="stockCount"
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="A brief description"
                label="Description"
                multiline={true}
                backgroundColor={COLORS.lightWhite}
                error={errors.description?.message}
              />
            )}
            name="description"
          />
        </View>
      </ScrollView>
      <View
        style={{
          ...styles.btnContainer,
        }}>
        <CustomButton title="Submit" onPress={handleSubmit(onSubmit)} />
      </View>
    </>
  );
}

const styles = StyleSheet.create<Style>({
  formContainer: {
    paddingTop: SPACING.small,
    paddingHorizontal: SPACING.xsmall,
  },
  btnContainer: {
    paddingBottom: SPACING.xsmall,
    paddingHorizontal: SPACING.xsmall,
  },
});

export default AddInventory;
