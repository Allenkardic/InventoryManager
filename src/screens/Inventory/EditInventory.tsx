import React from 'react';
import {Alert, ScrollView, StyleSheet, TextStyle, View} from 'react-native';
// components
import NavigationHeader from '../../components/NavigationHeader';
import CustomInput from '../../components/CustomInput';
import CustomButton from '../../components/CustomButton';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import {StackModels} from '../../navigations/models';
// context
import {AuthUserContext} from '../../context/AuthUserContext';
import {InventoryListContext} from '../../context/InventoryContext';

import {checkWordsCount} from '../../utils/constants';
import {COLORS, SPACING} from '../../utils/themes';
import type {InventoryCardProps} from '../../utils/types';
import {showMessage} from 'react-native-flash-message';
// form
import {useForm, Controller} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import * as yup from 'yup';
// @ts-ignore
import Icon from 'react-native-vector-icons/Feather';

type Style = {
  formContainer: TextStyle;
  btnContainer: TextStyle;
};

// navigation type
type Props = NativeStackScreenProps<StackModels, 'EditInventory'>;

// form type
type FormValues = {
  inventoryName: string;
  stockCount: number | string;
  description: string;
  price: number | string;
};

function EditInventory({navigation, route}: Props) {
  const screenData = route.params?.screenData;

  // states
  const [descriptionValue, setDescriptionValue] = React.useState('');

  const ss = 'dd hhh';

  console.log(checkWordsCount(ss), 'jj');
  // checkWordsCount(descriptionValue)
  // validation schema
  const schema = yup.object().shape({
    isCompany: yup.boolean(),
    // isCompany: yup
    //   .string()
    //   .test(
    //     'is boolean',
    //     'Please enter either 1 or 0',
    //     value => value === 'yes' || value === 'no',
    //   ),
    inventoryName: yup.string().required('Inventory name is required').trim(),
    stockCount: yup.number().required('Stock count left is required'),
    // description: yup
    //   .string()
    //   .required('Brief descripting is required')
    //   .when(yup.ref('$ss'),{
    //     is: true,
    //     then: yup.string().required('required'),
    //     otherwise: yup.string()
    // })
    //   }),

    description: yup
      .string()
      .test(
        'description',
        'Description not be less than 3 words',
        (value: any) => checkWordsCount(value) >= 3,
      ),
    // yup.string().required('minimum description is 3').mixed().oneOf([2]).isValid(2)

    // yup.string().required('Brief descripting is required'),

    // description: yup.array().min(3, 'hellow'),

    // .min(3, ({min}) => `Description must be greater than 3 wor`),
    price: yup.number().required('Price is required'),
  });

  // React Hook Form
  const {
    control,
    handleSubmit,
    formState: {errors},
  } = useForm<FormValues>({
    resolver: yupResolver(schema),
  });

  const {authUserStore} = React.useContext(AuthUserContext);
  const {inventoryStore, setInventoryStore} =
    React.useContext(InventoryListContext);

  const onSubmit = (data: FormValues) => {
    // update the product to waiting edit
    const updatedInventory = [...inventoryStore].map((el: any) => {
      if (el.id === screenData?.id) {
        el.inventoryName = data.inventoryName;
        el.price = data.price;
        el.stockCount = data.stockCount;
        el.description = data.description;
      }
      return el;
    });

    setInventoryStore(updatedInventory);

    showMessage({
      type: 'success',
      message: 'Item deleted successfuly',
    });
    navigation.navigate('Inventory');
  };

  const handleConfirmDelete = () => {
    const updatedInventory = inventoryStore.filter(
      el => el.id !== screenData?.id,
    );
    setInventoryStore(updatedInventory);

    showMessage({
      type: 'success',
      message: 'Item updated successfuly',
    });
    navigation.navigate('Inventory');
  };

  const handleDeletePress = () => {
    Alert.alert(
      `Confirm deleting ${screenData?.inventoryName}`,
      `Are you sure you want to delete ${screenData?.inventoryName}`,
      [
        {
          text: 'Cancel',
          onPress: () => {},
          style: 'cancel',
        },
        {text: 'OK', onPress: handleConfirmDelete},
      ],
    );
  };

  // const checkWordsCount = (word: string) => {
  //   const tremWords = word.trim().split(' ');
  //   return tremWords.length;
  // };

  // console.log(checkWordsCount(descriptionValue).length);
  // console.log(descriptionValue, 'hhh');
  return (
    <>
      <View>
        <NavigationHeader
          onPressBackBtn={() => navigation.goBack()}
          title={'Edit Inventory'}
          backgroundColor={COLORS.smokeWhite}
        />
      </View>
      <ScrollView>
        <View
          style={{
            ...styles.formContainer,
          }}>
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: book"
                label="Item name"
                backgroundColor={COLORS.lightWhite}
                error={errors.inventoryName?.message}
                marginBottom={SPACING.xsmall}
              />
            )}
            name="inventoryName"
            defaultValue={screenData?.inventoryName}
          />
          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 200"
                label="Price"
                backgroundColor={COLORS.lightWhite}
                error={errors.price?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="price"
            defaultValue={screenData?.price.toString()}
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={onChange}
                value={value}
                placeholder="ex: 10"
                label="Count"
                backgroundColor={COLORS.lightWhite}
                error={errors.stockCount?.message}
                marginBottom={SPACING.xsmall}
                keyboardType={'number-pad'}
              />
            )}
            name="stockCount"
            defaultValue={screenData?.stockCount.toString()}
          />

          <Controller
            control={control}
            render={({field: {onChange, onBlur, value}}) => (
              <CustomInput
                onChangeText={(text: string) => {
                  onChange(text);
                  setDescriptionValue(text);
                }}
                value={value}
                placeholder="A brief description"
                label="Description"
                multiline={true}
                backgroundColor={COLORS.lightWhite}
                error={errors.description?.message}
              />
            )}
            name="description"
            defaultValue={screenData?.description}
          />
        </View>
      </ScrollView>
      <View
        style={{
          ...styles.btnContainer,
        }}>
        <View style={{width: '45%'}}>
          <CustomButton title="Update" onPress={handleSubmit(onSubmit)} />
        </View>
        <View style={{width: '45%'}}>
          <CustomButton
            btnBackgroundColor={COLORS.red}
            title="Delete"
            onPress={handleDeletePress}
          />
        </View>
      </View>
    </>
  );
}

const styles = StyleSheet.create<Style>({
  formContainer: {
    paddingTop: SPACING.small,
    paddingHorizontal: SPACING.xsmall,
  },
  btnContainer: {
    paddingBottom: SPACING.xsmall,
    paddingHorizontal: SPACING.xsmall,
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
});

export default EditInventory;
