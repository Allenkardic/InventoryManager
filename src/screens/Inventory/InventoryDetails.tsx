import React from 'react';
import {StyleSheet, Text, View} from 'react-native';

import CustomText from '../../components/CustomText';
import CustomInput from '../../components/CustomInput';
// @ts-ignore
import Icon from 'react-native-vector-icons/Feather';

function InventoryDetails() {
  return (
    <View style={{paddingTop: 50}}>
      <CustomText>hello inventory details</CustomText>
      <CustomText>hello inventory details</CustomText>
      <CustomText>hello inventory details</CustomText>
      <CustomText>hello inventory details</CustomText>
    </View>
  );
}

export default InventoryDetails;
