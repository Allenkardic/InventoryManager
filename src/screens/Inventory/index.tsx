import React, {RefObject} from 'react';
import {FlatList, View, ListRenderItemInfo, StatusBar} from 'react-native';
// components
import InventoryCard from '../../components/InventoryCard';
import PlusButton from '../../components/PlusButton';
import NavigationHeader from '../../components/NavigationHeader';
import EmptyList from '../../components/EmptyList';

import {useForm} from 'react-hook-form';
import {yupResolver} from '@hookform/resolvers/yup';
import * as yup from 'yup';
import {NativeStackScreenProps} from '@react-navigation/native-stack';
import type {StackModels} from '../../navigations/models';
import type {InventoryCardProps} from '../../utils/types';
import {COLORS, HP, WP, SPACING, IMAGES} from '../../utils/themes';

// context
import {InventoryListContext} from '../../context/InventoryContext';
import {AuthUserContext} from '../../context/AuthUserContext';

// @ts-ignore
import Icon from 'react-native-vector-icons/Feather';
type Props = NativeStackScreenProps<StackModels, 'Inventory'>;

function Inventory({navigation}: Props) {
  // states
  const [inventoryList, setInventoryList] = React.useState<any>([]);
  // context
  const {authUserStore, setAuthUserStore} = React.useContext(AuthUserContext);
  const {inventoryStore} = React.useContext(InventoryListContext);

  React.useEffect(() => {
    // this checks and returns inventory list based on the sign in users email
    const inventoryByUser = inventoryStore.filter(
      item => item.userEmail === authUserStore.email,
    );
    setInventoryList(inventoryByUser);
  }, [inventoryStore]);

  const handleLogout = () => {
    setAuthUserStore({});
  };

  const handleOnPresscard = (item: InventoryCardProps) => {
    const screenData: InventoryCardProps = item;
    navigation.navigate('EditInventory', {screenData});
  };

  const renderItem = ({item}: ListRenderItemInfo<InventoryCardProps>) => {
    const {description, stockCount, price, inventoryName} = item;
    return (
      <View style={{marginHorizontal: SPACING.xxsmall}}>
        <InventoryCard
          onPress={() => handleOnPresscard(item)}
          inventoryName={inventoryName}
          stockCount={stockCount}
          price={price}
          description={description}
        />
      </View>
    );
  };

  return (
    <View style={{backgroundColor: COLORS.smokeWhite, flex: 1}}>
      <StatusBar
        animated={true}
        backgroundColor={COLORS.smokeWhite}
        barStyle="dark-content"
      />
      <NavigationHeader
        onPressRightIcon={handleLogout}
        rightIconName={'log-out'}
        title={'Inventory'}
        backgroundColor={COLORS.smokeWhite}
      />
      <View style={{marginTop: SPACING.xsmall}} />
      <FlatList
        data={inventoryList}
        renderItem={renderItem}
        keyExtractor={item => item.id?.toString() ?? ''}
        ListEmptyComponent={<EmptyList />}
      />

      <PlusButton onPress={() => navigation.navigate('AddInventory')} />
    </View>
  );
}

export default Inventory;
