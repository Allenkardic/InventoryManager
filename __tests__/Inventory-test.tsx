/**
 * @format
 */

import 'react-native';
import React from 'react';
import Inventory from '../src/screens/Inventory';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const createTestProps = (props: Object) => ({
  navigation: {
    navigate: jest.fn(),
  },
  ...props,
});

describe('Inventory Screen uses props', () => {
  it('renders correctly', () => {
    let props: any;
    props: createTestProps;
    const tree = renderer.create(<Inventory {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
