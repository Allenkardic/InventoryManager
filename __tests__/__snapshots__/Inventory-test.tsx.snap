// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Inventory Screen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#f5f5f5",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#f5f5f5",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 12,
        "paddingTop": 96,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "flex-start",
          "width": "10%",
        }
      }
    />
    <View
      style={
        Object {
          "width": "80%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Object {
            "alignContent": "center",
            "alignItems": "center",
            "fontSize": 31,
            "justifyContent": "center",
            "textAlign": "center",
          }
        }
      >
        Inventory
      </Text>
    </View>
    <View
      style={
        Object {
          "width": "10%",
        }
      }
    >
      <Text
        allowFontScaling={false}
        onPress={[Function]}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#121515",
              "fontSize": 40,
            },
            undefined,
            Object {
              "fontFamily": "Feather",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "marginTop": 48,
      }
    }
  />
  <RCTScrollView
    ListEmptyComponent={<Memo(EmptyList) />}
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "justifyContent": "center",
            "marginTop": 340,
          }
        }
      >
        <View
          style={
            Object {
              "height": 133.5,
              "width": 187.5,
            }
          }
        >
          <Image
            source={
              Object {
                "testUri": "../../../assets/images/empty-folder.png",
              }
            }
            style={
              Object {
                "height": "100%",
                "width": "100%",
              }
            }
          />
        </View>
        <Text
          allowFontScaling={false}
          style={
            Object {
              "alignContent": "center",
              "alignItems": "center",
              "fontSize": 20,
              "justifyContent": "center",
              "marginTop": 24,
              "textAlign": "center",
            }
          }
        >
          click the add icon to enter inventry
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#219653",
        "borderRadius": 46.75,
        "bottom": 80,
        "elevation": 5,
        "height": 93.5,
        "justifyContent": "center",
        "opacity": 1,
        "position": "absolute",
        "right": 52.5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 5,
          "width": 0,
        },
        "shadowOpacity": 0.2,
        "shadowRadius": 2,
        "width": 93.5,
        "zIndex": 2,
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        Array [
          Object {
            "color": "#ffffff",
            "fontSize": 53.5,
          },
          undefined,
          Object {
            "fontFamily": "Feather",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          Object {},
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;
